transport.go

```diff
@@ -38,77 +37,71 @@ func (t *Transport) RoundTrip(req *http.Request) (resp *http.Response, err error
 	span.SetStatus(semconv.SpanStatusFromHTTPStatusCode(resp.StatusCode))
+	resp, span = addResponseDataToSpanAndWrap(resp, span)
+
+	resp.Body = &wrappedBody{ctx: traceCtx, span: span, body: resp.Body}
+	return resp, err
+}
+
+func addRequestDataToSpanAndWrap(req *http.Request, span trace.Span) (*http.Request, trace.Span) {
+	req.Body, span = addBodyToSpan(req.Body, span, "http.request_body")
+	span = addHeaderToSpan(req.Header, span, "http.request_headers")
+	return req, span
+}
+
+func addResponseDataToSpanAndWrap(resp *http.Response, span trace.Span) (*http.Response, trace.Span) {
+	resp.Body, span = addBodyToSpan(resp.Body, span, "http.response_body")
+	span = addHeaderToSpan(resp.Header, span, "http.response_headers")
+	return resp, span
+}
+
+func addBodyToSpan(body io.ReadCloser, span trace.Span, attributeKey string) (io.ReadCloser, trace.Span) {
+	if body != nil {
+		bodyStr, bodyReadCloser, bodyErr := getFirstNCharsFromReadCloser(body, cfg.MaxEntrySize)
+		if bodyErr != nil {
+			logger.WithError(bodyErr).Error("failed to parse response body")
+		}
+		span.SetAttributes(attribute.String(attributeKey, bodyStr))
+		body = bodyReadCloser
+	}
+	return body, span
+}
 
-	responseHeaders := make(map[string]string)
-	for k, values := range resp.Header {
+func addHeaderToSpan(srcHeaders http.Header, span trace.Span, attributeKey string) trace.Span {
+	headers := make(map[string]string)
+	for k, values := range srcHeaders {
 		for _, value := range values {
-			responseHeaders[k] = value
+			headers[k] = value
 		}
 	}
-	headersJson, jsonErr := json.Marshal(responseHeaders)
+	headersJson, jsonErr := json.Marshal(headers)
 	if jsonErr != nil {
-		logger.WithError(err).Error("failed to fetch response headers")
+		logger.WithError(jsonErr).Error("failed to fetch request headers")
 	}
 	if len(headersJson) > cfg.MaxEntrySize {
-		span.SetAttributes(attribute.String("http.response_headers", string(headersJson[:cfg.MaxEntrySize])))
+		span.SetAttributes(attribute.String(attributeKey, string(headersJson[:cfg.MaxEntrySize])))
 	} else {
-		span.SetAttributes(attribute.String("http.response_headers", string(headersJson)))
+		span.SetAttributes(attribute.String(attributeKey, string(headersJson)))
 	}
+	return span
+}
 
-	if resp.Body != nil {
-		bodyBytes, bodyErr := io.ReadAll(resp.Body)
-		if bodyErr != nil {
-			logger.WithError(bodyErr).Error("failed to parse response body")
-		}
-		if len(bodyBytes) > cfg.MaxEntrySize {
-			span.SetAttributes(attribute.String("http.response_body", string(bodyBytes[:cfg.MaxEntrySize])))
-		} else {
-			span.SetAttributes(attribute.String("http.response_body", string(bodyBytes)))
-		}
-
-		resp.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
+func getFirstNCharsFromReadCloser(rc io.ReadCloser, n int) (string, io.ReadCloser, error) {
+	buf := make([]byte, n)
+	n, err := rc.Read(buf)
+	if err == io.EOF {
+		return string(buf), io.NopCloser(bytes.NewReader(buf[:n])), nil
+	} else if err != nil {
+		return "", nil, err
```

Comment:
In this case we will "destroy" the data that the user expects to get, isn't it?

Reply:
no. in this case the response was smaller than 2048 bytes we return the response without the padding. there is another bug here this should be:
return string(**buf[:n]**), io.NopCloser(bytes.NewReader(buf[:n])), nil